src/core/trace_core.mli

```diff
@@ -11,43 +12,60 @@ val enabled : unit -> bool
     This is fast, so that the traced program can check it before creating
     any span or message *)
 
-val enter_span :
+val with_span :
   ?__FUNCTION__:string ->
   __FILE__:string ->
   __LINE__:int ->
   ?data:(unit -> (string * user_data) list) ->
   string ->
-  span
-(** Enter a span. A span is a delimited period of time with
-    a start instant ("enter") and stop instant ("exit"), associated
-    with some metadata about the code entering/exiting some piece of code.
-
-    In particular the entrypoint comes with the location of the code
-    (you can use [__FILE__] and [__LINE__] directly in OCaml),
-    a mandatory name, and some optional metadata in a JSON-like representation.
-*)
+  (span -> 'a) ->
+  'a
+(** [with_span ~__FILE__ ~__LINE__ name f] enters a new span [sp],
+    and calls [f sp].
+    [sp] might be a dummy span if no collector is installed.
+    When [f sp] returns or raises, the span [sp] is exited.
 
-val exit_span : span -> unit
-(** Exit the span.
+    This is the recommended way to instrument most code.
 
-    This should be called exactly once per span (even in case of exception).
-    Once this is called, a timestamp might be recorded,
-  and the span is considered finished and can't be used anymore *)
+    {b NOTE} an important restriction is that this is only supposed to
+    work for synchronous, direct style code. Monadic concurrency, Effect-based
+    fibers, etc. might not play well with this style of spans on some
+    or all backends. If you use cooperative concurrency,
+    see {!enter_manual_span}.
```

Comment:
I'd definitely rewrite this; but it's a subtle point, and I'm not sure _exactly_ what to say instead.

Now that the API has been redesigned, `enter_explicit_span` and `with_span` should both work perfectly fine in both direct-style/on-the-stack _and_ promise/async-style/event-loop code just fine! The old 'uh-oh, implicit parent-tracking is entirely borked if you're in an event-loop' is _fixed_ now, specifically _because_ we add these functions with non-optional `parent` demands surfaced to the user. They "know" (in the sense of 'are forced to deal with') the shortcomings of the storage options if they choose functions that take explicit arguments, that they then must expose to their users. … right? :x

Really the only major caveat at this point, I think? unless I've become bogged down? is that `enter_initial_span` should prolly not be used by "libraries", only by top-level-apps, if you  want to be a good citizen.